content/blog/2023-12-11-lowering/index.md

```diff
@@ -0,0 +1,175 @@
+
+- For each temporary variable that X reads from, we must include loads before the instruction to load these variables from their stack locations  
+- For the temporary variable that X writes to, we must include a store after the instruction to store this variable to its appropriate stack location
+
+These conditions led us to implement functions for each RISCV instruction object - one to return the list of abstract temps that the instruction reads from, and one to return the list of abstract temps that the instruction writes to. Knowing these for each instruction would allow us to determine where to place surrounding instructions to shuttle temps on and off the stack.
+
+Now, we iterate through each RISCV instruction and get the list of abstract temps that it reads from and writes to. For each of the abstract temps that the instruction reads from, we use our mapper to look through the dictionary and find the stack offsets for each. Before this instruction, we add loads from these stack offsets into our RISCV registers. Then we perform the operation in the instruction, being sure to replace the abstract temps that are read from with the RISCV registers that we just loaded into. Now that the abstract temps in the instruction that we read from are taken care of, we move on to writes. We look at the abstract temp we write to, and replace that in the instruction with the next available RISCV register we haven't used for the reads yet. We then find using the mapper the stack location of this destination variable and we add an instruction after the operation that performs a store from the destination RISCV register into the appropriate stack offset the mapper gave us.
+
+Doing this for each RISCV instruction gives us executable assembly that is complete without any abstract temps besides dealing with function calls/arguments and calling conventions. 
+
+## Calling Conventions
+
+By far, the biggest implementation obstacle was implementing the RISC-V calling conventions. We based our implementation off of the RISC-V calling conventions taught in Cornell's CS 3410: Computer System Organization and Programming. To oversimplify four lectures' worth of material, the essence of RISC-V calling conventions can be distilled into three main phases:
+
+1. **Prologue (Function Entry Setup):** Before entering the function body, a set of preparatory actions, known as the prologue, take place. These include the creation of the stack frame. The size of the stack frame usually needs to be known ahead of time, considering the space needed to accommodate the return address, frame pointer, any overflow arguments, and local variables that must be pushed onto the stack. Additionally, all callee-save registers that will be utilized during the function execution should be pushed onto the stack at this stage.
```

Comment:
This chunk of the blog post might be a little clearer if you phrased things in terms of what you _did_, as opposed to what (abstractly) needs to be done for any implementation of function calls. As it stands, it's hard to tell what you have actually implemented, versus what is still "to do." For example, you say:

> The size of the stack frame usually needs to be known ahead of time

Is this true in your compiler? Did you calculate the size of the stack frame?

Reply:
That makes sense - I've added discussion on what we actually did in our implementation in the Prologue, FunctionCall, and Epilogue sub-sections.